beacon-chain/sync/rpc_chunked_response.go

```diff
@@ -39,48 +42,77 @@ func ReadChunkedBlock(stream libp2pcore.Stream, chain blockchain.ChainInfoFetche
 	if isFirstChunk {
 		return readFirstChunkedBlock(stream, chain, p2p)
 	}
-	blk := &eth.SignedBeaconBlock{}
-	if err := readResponseChunk(stream, chain, p2p, blk); err != nil {
-		return nil, err
-	}
-	return interfaces.WrappedPhase0SignedBeaconBlock(blk), nil
+	return readResponseChunk(stream, chain, p2p)
 }
 
 // readFirstChunkedBlock reads the first chunked block and applies the appropriate deadlines to
 // it.
 func readFirstChunkedBlock(stream libp2pcore.Stream, chain blockchain.ChainInfoFetcher, p2p p2p.P2P) (interfaces.SignedBeaconBlock, error) {
-	blk := &eth.SignedBeaconBlock{}
 	code, errMsg, err := ReadStatusCode(stream, p2p.Encoding())
 	if err != nil {
 		return nil, err
 	}
 	if code != 0 {
 		return nil, errors.New(errMsg)
 	}
-	// No-op for now with the rpc context.
-	_, err = readContextFromStream(stream, chain)
+	rpcCtx, err := readContextFromStream(stream, chain)
 	if err != nil {
 		return nil, err
 	}
+	blk, err := extractBlockDataType(rpcCtx, chain)
+	if err != nil {
+		return nil, err
+	}
+	// This may not work, double check tests.
 	err = p2p.Encoding().DecodeWithMaxLength(stream, blk)
-	return interfaces.WrappedPhase0SignedBeaconBlock(blk), err
+	return blk, err
 }
 
 // readResponseChunk reads the response from the stream and decodes it into the
 // provided message type.
-func readResponseChunk(stream libp2pcore.Stream, chain blockchain.ChainInfoFetcher, p2p p2p.P2P, to interface{}) error {
+func readResponseChunk(stream libp2pcore.Stream, chain blockchain.ChainInfoFetcher, p2p p2p.P2P) (interfaces.SignedBeaconBlock, error) {
 	SetStreamReadDeadline(stream, respTimeout)
 	code, errMsg, err := readStatusCodeNoDeadline(stream, p2p.Encoding())
 	if err != nil {
-		return err
+		return nil, err
 	}
 	if code != 0 {
-		return errors.New(errMsg)
+		return nil, errors.New(errMsg)
 	}
 	// No-op for now with the rpc context.
-	_, err = readContextFromStream(stream, chain)
+	rpcCtx, err := readContextFromStream(stream, chain)
 	if err != nil {
-		return err
+		return nil, err
+	}
+	blk, err := extractBlockDataType(rpcCtx, chain)
+	if err != nil {
+		return nil, err
+	}
+	// This may not work, double check tests.
+	err = p2p.Encoding().DecodeWithMaxLength(stream, blk)
+	return blk, err
+}
+
+func extractBlockDataType(digest []byte, chain blockchain.ChainInfoFetcher) (interfaces.SignedBeaconBlock, error) {
+	if len(digest) == 0 {
+		bFunc, ok := types.BlockMap[bytesutil.ToBytes4(params.BeaconConfig().GenesisForkVersion)]
+		if !ok {
+			return nil, errors.New("no block type exists for the genesis fork version.")
+		}
+		return bFunc(), nil
+	}
+	if len(digest) != contextLength {
+		return nil, errors.Errorf("invalid digest returned, wanted a length of %d but received %d", contextLength, len(digest))
+	}
+	vRoot := chain.GenesisValidatorRoot()
+	for k, blkFunc := range types.BlockMap {
```

Comment:
Given the map structure, I think we can just do a O(1) look up instead of O(n) looping through the map items to find the match.  One way is to use `rDigest` as the key